python/route_guide/route_guide_server.py

```diff
@@ -0,0 +1,147 @@
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""The Python implementation of the gRPC route guide server."""
+
+import json
+import time
+from math import sqrt, radians, sin, cos, atan2
+
+import route_guide_pb2
+
+_ONE_DAY_IN_SECONDS = 60 * 60 * 24
+
+
+def read_route_guide_db():
+  """Reads the route guide"""
+  db = []
+  with open("route_guide_db.json") as f:
+    for item in json.load(f):
+      feature = route_guide_pb2.Feature(name = item["name"])
+      feature.location.longitude = item["location"]["longitude"]
+      feature.location.latitude = item["location"]["latitude"]
+      db.append(feature)
+  return db
+
+
+def get_feature(feature_db, point):
+  """Returns feature at given location or None"""
+  for feature in feature_db:
+    if feature.location.latitude == point.latitude and feature.location.longitude == point.longitude:
+      return feature
+  return None
+
+
+def get_distance(start, end):
+  """Distance between two points"""
+  coord_factor = 10000000.0
+  lat_1 = start.latitude / coord_factor
+  lat_2 = end.latitude / coord_factor
+  lon_1 = start.latitude / coord_factor
+  lon_2 = end.longitude / coord_factor
+  lat_rad_1 = radians(lat_1)
+  lat_rad_2 = radians(lat_2)
+  delta_lat_rad = radians(lat_2 - lat_1)
+  delta_lon_rad = radians(lon_2 - lon_1)
+
+  a = pow(sin(delta_lat_rad/2), 2) + cos(lat_rad_1) * cos(lat_rad_2) * pow(sin(delta_lon_rad/2), 2)
+  c = 2 * atan2(sqrt(a), sqrt(1-a))
+  R = 6371000; # metres
+  return R * c;
+
+class RouteGuideServicer(route_guide_pb2.EarlyAdopterRouteGuideServicer):
+  """Provides methods that implement functionality of route guide server."""
+
+  def __init__(self):
+      self.db = read_route_guide_db()
+
+  def GetFeature(self, request, context):
+    feature = get_feature(self.db, request)
+    if not feature:
+      feature = route_guide_pb2.Feature(name = "")
+      feature.location.longitude = request.longitude
+      feature.location.latitude = request.latitude
+    return feature
+
+  def ListFeatures(self, request, context):
+    lo = request.lo
+    hi = request.hi
+    left = min(lo.longitude, hi.longitude)
+    right = max(lo.longitude, hi.longitude)
+    top = max(lo.latitude, hi.latitude)
+    bottom = min(lo.latitude, hi.latitude)
+    for f in self.db:
```

Comment:
Rename "f" to "feature". Many Python programmers use "f" for file objects so often that it misleads them to see "f" stand for anything else. :-P